keanu-project/src/main/java/io/improbable/keanu/algorithms/graphtraversal/DifferentiableChecker.java

```diff
@@ -0,0 +1,90 @@
+package io.improbable.keanu.algorithms.graphtraversal;
+
+import io.improbable.keanu.vertices.Vertex;
+import lombok.experimental.UtilityClass;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Set;
+
+@UtilityClass
+public class DifferentiableChecker {
+
+    /**
+     * Checks whether a set of vertices are differentiable
+     * w.r.t the latents in a graph.
+     *
+     * @param vertices the collection of vertices to check
+     * @return true if the vertices are differentiable w.r.t latents
+     */
+    public boolean isDifferentiable(Collection<Vertex> vertices) {
+        Queue<Vertex> queue = new LinkedList<>(vertices);
+        Set<Vertex> queued = new HashSet<>(vertices);
+        Set<Vertex> constantValueVerticesCache = new HashSet<>();
```

Comment:
I think "cache" is a misleading name here as it suggests it's just an optimisation and doesn't affect the functionality.

Reply:
Have made some changes to the algorithm. It now is just an optimisation.